firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,148 @@
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new WeakHashMap<>(), false);
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(Activity activity, FrameMetricsAggregator frameMetricsAggregator, WeakHashMap<Object, PerfFrameMetrics> subTraceMap, boolean isRecording) {
+    this.activity = activity;
+    this.fma = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+    this.isRecording = isRecording;
+  }
+
+  /**
+   * Starts recording FrameMetrics for the activity window
+   */
+  public void start() {
+    if (isRecording) {
+      logger.error("FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    fma.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window
+   * @return FrameMetrics accumulated during the current recording
+   */
+  public PerfFrameMetrics stop() {
+    if (!isRecording) {
+      logger.error("Cannot stop because no recording was started");
+      return null;
+    }
+    isRecording = false;
+    PerfFrameMetrics data = this.snapshot();
+    try {
+      fma.remove(activity);
+    } catch (IllegalArgumentException ignored) {
+      logger.error("View not hardware accelerated. Unable to collect FrameMetrics.");
+    }
+    fma.reset();
+    return data;
+  }
+
+  /**
+   * Starts a sub-trace in the current recording
+   * @param key a UI state to associate this sub-trace with. e.g.) fragment
+   */
+  public void startSubTrace(Object key) {
+    if (!isRecording) {
+      logger.error("Cannot start sub-trace because FrameMetricsAggregator is not recording");
+      return;
+    }
+    if (subTraceMap.containsKey(key)) {
+      logger.error("Cannot start sub-trace because one is already ongoing with the key %s", key.getClass().getSimpleName());
+      return;
+    }
+    subTraceMap.put(key, this.snapshot());
+  }
+
+  /**
+   * Stops the sub-trace associated with the given UI state
+   * @param key the UI state associated with the sub-trace to be stopped
+   * @return FrameMetrics accumulated during this sub-trace
+   */
+  public PerfFrameMetrics stopSubTrace(Object key) {
+    if (!isRecording) {
+      logger.error("Cannot stop sub-trace because FrameMetricsAggregator is not recording");
+      return null;
+    }
+    if (!subTraceMap.containsKey(key)) {
+      logger.error("Sub-trace associated with key %s was not started or does not exist", key.getClass().getSimpleName());
+      return null;
+    }
+    PerfFrameMetrics snapshotStart = subTraceMap.remove(key);
+    PerfFrameMetrics snapshotEnd = this.snapshot();
+    return snapshotEnd.subtract(snapshotStart);
+  }
+
+  /**
+   * Calculate total frames, slow frames, and frozen frames from SparseIntArray[] recorded by {@link
+   * FrameMetricsAggregator}.
+   *
+   * @return the frame metrics
+   */
+  protected @NonNull PerfFrameMetrics snapshot() {
```

Comment:
Yes, I copied that method over here because I realize no other class will use that method once we encapsulate everything related to FMA, including calculations, in the Recorder. I think it's better to make that into a implementation detail hidden away by the recorder (private/protected method). 

I was planning to remove Calculator and make a separate file for `PerfFrameMetrics` later in the 2nd PR, because `AppStateMonitor` and `FragmentStateMonitor` uses the Calculator, and we still need `PerfFrameMetrics`.

Reply:
On second thought, the recorder class is getting too big and too difficult to test if this is moved over, so reverted it and now `snapshot()` uses `FrameMetricsCalculator`.